src/modules/mavlink/mavlink_log_handler.cpp

```diff
@@ -31,558 +31,498 @@
-						strncpy(filename, file, filename_len);
-						filename[filename_len - 1] = 0; // ensure null-termination
-					}
-
-					result = true;
-					break;
-				}
-			}
-		}
+	mavlink_log_request_list_t request;
+	mavlink_msg_log_request_list_decode(msg, &request);
 
-		fclose(f);
+	if (!create_log_list_file()) {
+		return;
 	}
 
-	return result;
+	_list_request.first_id = request.start;
+	_list_request.last_id = request.end == 0xffff ? _num_logs : request.end;
+	_list_request.current_id = 0;
+	_list_request.file_index = 0;
+	_logs_listed = true;
+	_state = LogHandlerState::Listing;
 }
 
-//-------------------------------------------------------------------
-bool
-MavlinkLogHandler::_open_for_transmit()
+void MavlinkLogHandler::handle_log_request_data(const mavlink_message_t *msg)
 {
-	if (_current_log_filep) {
-		::fclose(_current_log_filep);
-		_current_log_filep = nullptr;
+	if (!_logs_listed) {
+		PX4_WARN("Logs not yet listed");
+		_state = LogHandlerState::Idle;
+		return;
 	}
 
-	_current_log_filep = ::fopen(_current_log_filename, "rb");
+	mavlink_log_request_data_t request;
+	mavlink_msg_log_request_data_decode(msg, &request);
 
-	if (!_current_log_filep) {
-		PX4LOG_WARN("MavlinkLogHandler::open_for_transmit Could not open %s", _current_log_filename);
-		return false;
+	if (request.id >= _num_logs) {
+		PX4_WARN("Requested log %" PRIu16 " but we only have %u", request.id, _num_logs);
+		_state = LogHandlerState::Idle;
+		return;
 	}
 
-	return true;
-}
+	// Handle switching to new request ID
+	if (request.id != _current_entry.id) {
+		// Close the old file
+		if (_current_entry.fp) {
+			fclose(_current_entry.fp);
+			_current_entry.fp = nullptr;
+		}
 
-//-------------------------------------------------------------------
-size_t
-MavlinkLogHandler::_get_log_data(uint8_t len, uint8_t *buffer)
-{
-	if (!_current_log_filename[0]) {
-		return 0;
-	}
+		LogEntry entry = {};
 
-	if (!_current_log_filep) {
-		PX4LOG_WARN("MavlinkLogHandler::get_log_data file not open %s", _current_log_filename);
-		return 0;
-	}
+		if (!log_entry_from_id(request.id, &entry)) {
+			PX4_WARN("Log file ID %u does not exist", request.id);
+			_state = LogHandlerState::Idle;
+			return;
+		}
 
-	long int offset = _current_log_data_offset - ftell(_current_log_filep);
+		entry.fp = fopen(entry.filepath, "rb");
+		entry.offset = request.ofs;
 
-	if (offset && fseek(_current_log_filep, offset, SEEK_CUR)) {
-		fclose(_current_log_filep);
-		_current_log_filep = nullptr;
-		PX4LOG_WARN("MavlinkLogHandler::get_log_data Seek error in %s", _current_log_filename);
-		return 0;
+		if (!entry.fp) {
+			PX4_WARN("Failed to open file %s", entry.filepath);
+			return;
+		}
+
+		_current_entry = entry;
 	}
 
-	size_t result = fread(buffer, 1, len, _current_log_filep);
-	return result;
-}
+	// Stop if offset request is larger than file
+	if (request.ofs >= _current_entry.size_bytes) {
+		PX4_WARN("Request offset %" PRIu32 "greater than file size %" PRIu32, request.ofs,
```

Comment:
If you're going to keep these warnings enabled by default I'd do a pass to make sure they have enough context to be useful. For example mentioning it's the log handler, which log file if relevant, and any meaningful result (aborted? retrying?).